src/dataflow/src/server.rs

```diff
@@ -818,6 +1042,15 @@ where
             }
         }
     }
+
+    /// Attempt to compact source timestamp bindings
+    fn compact_timestamp_bindings(&self) {
+        for (_, ts_history) in self.render_state.ts_histories.borrow().iter() {
+            if let TimestampDataUpdate::BringYourOwn(history) = ts_history {
+                history.compact();
+            }
+        }
+    }
```

Comment:
Is there a reason this is not event-driven? It seems like compaction could be driven entirely on receipt of the events that allow it.

Reply:
no particular reason. i wanted to limit the number of times we compacted in case the coordinator was spammy i guess?

Reply:
re-upping this comment. we receive an event that allows compaction; why is that not the right moment to apply the compaction, vs rescanning all histories regularly?

Reply:
My main worry is that it is further abstraction leak. The server now needs to know about a method that has to be called with some frequency idk, vs allowing it to be handled automatically as part of the other event that it needs to pass along.

Reply:
On further reading, it seems like compaction can be made to take time proportional to the compaction performed, which would be fine to call with whatever frequency. However, at the moment each call performs linear work in the size of the history, rather than the compaction. This should probably be cleaned up in the future (I'd file an issue if you want to land this as is).